### ISSUE TITLE
jspm bundle / build --watch flag

### ISSUE THREAD

**Issue by guybedford** [jspm/jspm-cli#1572]
Would provide a bundle-based development workflow. Relatively simple to implement based on adding fs watching to the principles of https://github.com/jackfranklin/jspm-dev-builder if anyone is interested in providing a PR.

**Comment by AdamBurgess** [comment]
I tried to implement it: https://github.com/AdamBurgess/jspm-cli/commit/4f1032dfd75427d40ad1f6865b0def3e45d03a7f not asking for comments just yet, though it does work
initial startup + first build are slow, then all builds after are super fast, which is awesome
currently only builds are implemented. It'd be nice to have the build and bundle command doing mostly the same steps, so that definitely could be refactored

**Comment by guybedford** [comment]
Nice! Yeah it might be writing it as a higher level wrapper around the function to work against build and bundle and others in future (compile should really be here too) something like:

```javascript
watchCommand(function() {
  return build(x, x, x);
}, watchOptions);
```

Where `watchOptions` has things like `watchDir` and perhaps timeout options as well.

**Comment by guybedford** [comment]
`watch.js` could even be its own file then in the `lib` dir...

**Comment by benzen** [comment]
This could be simplly a node module for the sake of simplicity.

I understand that's not the best behavior, but we only wait for http2 to happens

**Comment by mgamsjager** [comment]
@benzen Why not use chokidar which is some what the industry standard.

**Comment by benzen** [comment]
I never went to see what is chokidar.
But i know a lot of project use it.
To me both looks fine

2016-03-14 9:21 GMT-04:00 Matthias Gamsjager <notifications@github.com>:

> @benzen <https://github.com/benzen> Why not use chokidar which is some
> what the industry standard.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jspm/jspm-cli/issues/1572#issuecomment-196307979>.
>



-- 
Benjamin DREUX

**Comment by guybedford** [comment]
The approach in https://github.com/oligot/jspm-dev-builder-cli seems ideal, and would be the exact way we should implement here, except that we should watch the config file itself as well for changes.

**Comment by born2net** [comment]
I personally don't like to debug on a bundle as you normally don't get error with file reference and source mapping is never as efficient as one debugging an un bundled app, its the reason I moved away webpack,

regards

Sean

**Comment by guybedford** [comment]
This has been implemented for `jspm bundle` in https://github.com/jspm/jspm-cli/commit/61a17963baf51c2cfb270ec10e6f55181aac2a19, pending release.

**Comment by guybedford** [comment]
A massive thank you to @oligot and @jackfranklin for spearheading this area and making this possible.

**Comment by mgamsjager** [comment]
Great you got it integrated so fast.

Any benchmarks yet?

**Comment by mgamsjager** [comment]
I kinda run into this error. Updated system with jspm dl-loader --edge but didn't made any difference
```
ok   Built into build.js with source maps, unminified.
     Watching src for changes...
ok   File boot.ts changed, rebuilding...

err  TypeError: mapped.substr is not a function
    at validateMapping (C:\dev\jspm\node_modules\systemjs\dist\system.src.js:2140:21)
    at doMapSync (C:\dev\jspm\node_modules\systemjs\dist\system.src.js:2147:5)
    at applyPackageConfigSync (C:\dev\jspm\node_modules\systemjs\dist\system.src.js:2128:16)
    at SystemJSNodeLoader.<anonymous> (C:\dev\jspm\node_modules\systemjs\dist\system.src.js:2338:14)
    at SystemJSNodeLoader.normalizeSync (C:\dev\jspm\node_modules\systemjs\dist\system.src.js:4238:30)
    at SystemJSNodeLoader.loader.normalizeSync (C:\dev\jspm\node_modules\systemjs-builder\lib\builder.js:223:42)
    at C:\dev\jspm\node_modules\systemjs-builder\lib\builder.js:92:29
    at Array.forEach (native)
    at Builder.invalidate (C:\dev\jspm\node_modules\systemjs-builder\lib\builder.js:91:27)
    at C:\dev\jspm\lib\bundle.js:186:25
    at NodeWatcher.<anonymous> (C:\dev\jspm\lib\bundle.js:291:5)
    at emitThree (events.js:97:13)
    at NodeWatcher.emit (events.js:175:7)
    at NodeWatcher.<anonymous> (C:\dev\jspm\node_modules\sane\src\node_watcher.js:315:10)
    at Timer.listOnTimeout (timers.js:92:15)

```

**Comment by guybedford** [comment]
@mgamsjager yes there was a builder patch added in https://github.com/systemjs/builder/commit/bb312f790fdd260ae329b2216ecb6c324a25026e as well to go with this release.

**Comment by mgamsjager** [comment]
@guybedford with this patch it works. And quite fast :)

**Comment by mgamsjager** [comment]
It seems to work only for existing files. When adding a new file (and importing it in another file), it is not part of the bundle.js

**Comment by guybedford** [comment]
@mgamsjager it should definitely work with that. If you can share the exact steps to replicate that, that would help.

**Comment by AdamBurgess** [comment]
is watch support for `jspm build` still on the table? or was there some issue preventing it from being implemented?

**Comment by guybedford** [comment]
The blocker for this is https://github.com/systemjs/builder/issues/399, which shouldn't be too tricky.

**Comment by mgamsjager** [comment]
Angular2 with Typescript
The command I ran: `node ..\jspm\jspm bundle src/boot.ts --watch` (got jspm repo in another folder)

boot.ts
```
// Global deps
import 'reflect-metadata';
import 'zone.js'; 

// Angular deps
import {bootstrap} from 'angular2/platform/browser';     

//import {Shell} from './components/shell/shell-component.ts';

bootstrap(
  //  Shell    
);

```
Build looks like this:
```
 npm:rxjs@5.0.0-beta.4/util/toSubscriber.js
 npm:rxjs@5.0.0-beta.4/util/tryCatch.js
 npm:zone.js@0.6.6.json
 npm:zone.js@0.6.6/dist/zone.js
 src/boot.ts
```

So when removing the // from the boot.ts while in watch
```
// Global deps
import 'reflect-metadata';
import 'zone.js'; 

// Angular deps
import {bootstrap} from 'angular2/platform/browser';     

import {Shell} from './components/shell/shell-component.ts';

bootstrap(
    Shell    
);
```

The build still look like this:

``` 
  npm:rxjs@5.0.0-beta.4/util/throwError.js
  npm:rxjs@5.0.0-beta.4/util/toSubscriber.js
  npm:rxjs@5.0.0-beta.4/util/tryCatch.js
  npm:zone.js@0.6.6.json
  npm:zone.js@0.6.6/dist/zone.js
  src/boot.ts

```

But when resetting the build it looks like it should:
```
 npm:rxjs@5.0.0-beta.4/util/toSubscriber.js
 npm:rxjs@5.0.0-beta.4/util/tryCatch.js
 npm:zone.js@0.6.6.json
 npm:zone.js@0.6.6/dist/zone.js
 src/boot.ts
 src/components/shell/shell-component.ts
```

Please let me know if you need more specifics

**Comment by guybedford** [comment]
@mgamsjager thanks for the detailed example here. I believe this was just fixed by @oligot in https://github.com/jspm/jspm-cli/pull/1656, but let me know if it still gives issues with that fix.

**Comment by mgamsjager** [comment]
@guybedford It's fixed

**Comment by born2net** [comment]
where can I learn more about what watch is for? 
does it work with source maps for debugging in ng2 / TS?
any docs?

tx

Sean

**Comment by oligot** [comment]
@born2net You can have a look at a first documentation draft here:
https://github.com/jspm/jspm.io/pull/43/files?short_path=4cea63c#diff-4cea63c9d39d3e90a68b25b0030e90aa

**Comment by guybedford** [comment]
Fixed in jspm 0.17.0-beta.12, with `jspm build --watch` support too.

**Comment by born2net** [comment]
great, will give it a shot...

**Comment by born2net** [comment]
in beta 0.12 I am getting an error 

```
jspm bundle src/App.js ./dist/index.js --skip-source-maps
     Building the bundle tree for src/App.js...

err  Error: ENOENT: no such file or directory, open 'C:\msweb\studioDashboard\css'
      at Error (native)
  Error loading C:/msweb/studioDashboard/css
```
I don't have or use a directory named C:/msweb/studioDashboard/css so I am not sure why the bundler is looking for that dir...

regards

Sean.

**Comment by onlywei** [comment]
@guybedford do I have to use the CLI like that, or is there an API I can call from JavaScript?

**Comment by guybedford** [comment]
@onlywei this is a CLI-only feature. The builder API is designed to chain with other tools rather (cache invalidation there is purely via `builder.invalidate('module/x.js')`).

**Comment by jackfranklin** [comment]
@guybedford this is awesome! Am I right to think this effectively deprecates jspm-dev-builder for those running jspm 0.17 ?

I'll update my repo to make people aware of that :+1:

**Comment by guybedford** [comment]
@jackfranklin thanks for your support! The dev builder may still be useful for the case @onlywei mentions for having a watcher for the builder API, which we don't extend this support into.

**Comment by jackfranklin** [comment]
Ah cool, so this is entirely CLI only? Understood. Would work on exposing the API that the jspm CLI uses be of interest? That's probably something I could work on / PR into jspm.

**Comment by kwesterfeld** [comment]
@guybedford I started using this feature and I LOVE it.  However, I found that because I have baseURL set in package.json, the `WatchmanWatcher#buildWatch` method dies with ENOENT:

```
jspm bundle 'app/main + systemjs-hot-reloader' smc/bundle/smc-bundle.js --inject --watch
```

Where "smc" is set as my baseURL in package.json.   The error:

```
err  Error: watch /Users/..../smc/ui/app ENOENT
    at exports._errnoException (util.js:856:11)
    at FSWatcher.start (fs.js:1313:19)
    at Object.fs.watch (fs.js:1341:11)
```

My workaround was to add a symlink in the directory where it's trying to find "app".  This would be PERFECT if the `WatchmanWatcher#buildWatch` method would prefix baseURL from package.json.

**Comment by guybedford** [comment]
@kwesterfeld thanks for the feedback here, this is fixed in https://github.com/jspm/jspm-cli/commit/bbaa11b65d569ab48ceb68e49baa3d668cb7db89.

**Comment by Simon-Li** [comment]
JSPM v0.17, quite expected, what is the release date of it? even roughly. thanks!

**Comment by sergei-maertens** [comment]
@jackfranklin have you gotten anywhere with exposing the API? I would love to be able to start a watch task in `gulp`

**Comment by ArmorDarks** [comment]
I wonder, is it possible somehow to use JSPM `--watch` with Browser Sync? Except spawning parallel process, of course...

**Comment by sergei-maertens** [comment]
I believe you're looking for the term 'live-reload' in the jspm beta docs

**Comment by ArmorDarks** [comment]
@sergei-maertens Did you mean http://jspm.io/0.17-beta-guide/hot-reloading.html? Well, it has same questions — it seems you need to spawn two parallel wathcing processes if you want to run hot-reloader and Browser Sync

**Comment by piotrwitek** [comment]
@ArmorDarks you cannot out-of-the-box, but it seems that they both use socket.io under the hood, so this is just a matter of hooking the events used by one of them for both and it probably should work fine

**Comment by ArmorDarks** [comment]
I see, thanks. Just wondering did someone already do that. Wasn't able to find working examples, so far.

**Comment by oligot** [comment]
[Here is an example](https://github.com/oligot/mithril-es6-modules/blob/9726a1dca2c73609064eba496503869b57cd8b5f/package.json#L6-L8) on how to run `jspm bundle/build --watch` and browser sync, using [npm-run-all](https://github.com/mysticatea/npm-run-all) to run the scripts in parallel (using the  `--print-label` option to print the task name as a prefix on each line of output).

We are using it quite extensively and it works quite well.

**Comment by ArmorDarks** [comment]
@oligot Thanks. That's exactly what I came to too, just wasn't sure it was right approach. So far it seems to be the only one possible

Is API for watch planned at all?.

**Comment by sergei-maertens** [comment]
No, I should change the title of this issue :-)

**Comment by jonaskello** [comment]
Same as @ArmorDarks I also need clarification on how the html should look like with the development bundle approach. I followed the [guide](http://jspm.io/0.17-beta-guide/development-bundling.html) but it does not give a new version of the `test.html` file from the beginning of the guide. Does anything need to change in test.html from the original version in the "Creating a project" section? I guess the html should not import test.js but bundle.js instead. I tried multiple perumuations of importing the scripts but none work.
